File: ts_gui/src/invoke/store.ts
import { IStoreProfiles } from "@/interfaces/IStoreProfiles";
import { IStoreSettings } from "@/interfaces/IStoreSettings";
import { IStoreSettingsJava } from "@/interfaces/IStoreSettingsJava";
import { invoke } from "@tauri-apps/api/tauri";

export enum JavaPathVersion {
    Java8 = 0,
    Java16 = 1,
    Java17 = 2
}

export default class Store {
    
    public static async saveLauncherSettingsFile(): Promise<void> {
        await invoke("plugin:store|save");
    }

    public static async getSettings(): Promise<IStoreSettings> {
        return await invoke("plugin:store|settings_get");
    }
  
    public static async setSettings(value: IStoreSettings) {
    public static async getSettingsSelectedServerStart(): Promise<IStoreSettingSelectedServer> {
        return await logger.invokeWithLogging<IStoreSettingSelectedServer>("plugin:store|get_settings_selected_server_start");
    }

    public static async setSettingsSelectedServerStart(value: IStoreSettingSelectedServer) {
        await logger.invokeWithLogging<void>("plugin:store|set_settings_selected_server_start", { value });
    }

    public static async getSettingsJava(id: string): Promise<IStoreSettingsJava> {
        return await invoke("plugin:store|get_settings_java", { id });
    }

    public static async setSettingsJava(id: string, value: IStoreSettingsJava): Promise<void> {
        await invoke("plugin:store|set_settings_java", { id, value });
    }

    public static async getProfiles(): Promise<IStoreProfiles> {
        return await invoke("plugin:store|profiles_get");
    }
  
    public static async setProfiles(value: IStoreProfiles) {
        await invoke("plugin:store|profiles_set", { value });
    }

    public static async clearProfiles() {
        await invoke("plugin:store|profiles_clear");
    }
}

File: ts_gui/src-tauri/src/api/store.rs
use allay_core::{data::{Java, Profiles, Settings}, profiles, settings, SelectedServerStart, Store};
use tauri::plugin::TauriPlugin;

pub fn init<R: tauri::Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::new("store")
        .invoke_handler(tauri::generate_handler![
            save,
            settings_get,
            settings_set,
            get_settings_java,
            set_settings_java,
            profiles_get,
            profiles_set,
            profiles_clear,
            get_settings_selected_server_start,
            set_settings_selected_server_start
        ])
        .build()
}

// invoke("plugin:store|save")
#[tauri::command]
pub async fn save() -> crate::api::Result<()> {
    Store::sync().await?;
    Ok(())
}

// invoke("plugin:store|settings_get")
#[tauri::command]
pub async fn settings_get() -> super::Result<Settings> {
    Ok(settings::get().await?)
}

// invoke("plugin:store|settings_set", settings)
#[tauri::command]
pub async fn settings_set(value: Settings) -> super::Result<()> {
    settings::set(value).await?;
    Ok(())
}

// invoke("plugin:store|get_settings_java", id)
#[tauri::command]
pub async fn get_settings_java(id: &str) -> super::Result<Java> {
    Ok(settings::get_java(id).await?)
}

// invoke("plugin:store|get_settings_java", { id, java })
#[tauri::command]
pub async fn set_settings_java(id: &str, value: Java) -> super::Result<()> {
    settings::set_java(id, value).await?;
    Ok(())
}

// invoke("plugin:store|profiles_get")
#[tauri::command]
pub async fn profiles_get() -> super::Result<Profiles> {
    Ok(profiles::get().await?)
}

// invoke("plugin:store|profiles_set", profiles)
#[tauri::command]
pub async fn profiles_set(value: Profiles) -> super::Result<()> {
    profiles::set(value).await?;
    Ok(())
}

// invoke("plugin:store|profiles_clear")
#[tauri::command]
pub async fn profiles_clear() -> super::Result<()> {
    profiles::clear().await?;
    Store::sync().await?;
    Ok(())
}

// invoke("plugin::store|get_settings_selected_server_start")
#[tauri::command]
pub async fn get_settings_selected_server_start() -> super::Result<SelectedServerStart>{
    Ok(settings::get_selected_server_start().await?)
}


// invoke("plugin::store|set_settings_selected_server_start")
#[tauri::command]
pub async fn set_settings_selected_server_start(value: SelectedServerStart) -> super::Result<()>{
    match settings::set_selected_server_start(value.clone()).await {
        Ok(_) => {
            tracing::debug!("Selected Server Start set: {:?}", value);
            match Store::sync().await {
                Ok(_) => {
                    tracing::debug!("Store synced successfully.");
                    Ok(())
                }
                Err(e) => {
                    tracing::error!("Failed to sync store: {:?}", e);
                    Err(e.into()) // Convert the error to your custom Result type
                }
            }
        }
        Err(e) => {
            tracing::error!("Failed to set selected server start: {:?}", e);
            Err(e.into()) // Convert the error to your custom Result type
        }
    }
    
    
    
    
    //settings::set_selected_server_start(value).await?;
    //tracing::debug!("Selected Server Start set:{:?}",value);
    //Store::sync().await?;
    //Ok(())
}
